Add Missing URL state to Navigation Link Block

georgeh opened the pull request:
## Description
Changing the Navigation Link block to highlight when a link is missing 

Closes #28440
Closes #23326

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

## Screenshots <!-- if applicable -->

https://user-images.githubusercontent.com/128240/108095707-28406180-704e-11eb-8d2a-981b6790a55c.mov

## Types of changes

When the URL is missing from a Navigation Link Block:
* Adds a placeholder
* Forces the URL popover open when the block is selected
* Prevents editing the text

Bugfix:
* Cancels any outstanding requests in `URLInput` when the component is unmounted to avoid calling `setState` on an unmounted component
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [na] ~~I've included developer documentation if appropriate.~~ <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [na] ~~I've updated all React Native files affected by any refactorings/renamings in this PR.~~ <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/native-mobile.md -->

Commit:
Prototyping a new Link Nav Block Interaction

Working on #28440

Co-authored-by: Kerry Liu <kerry.liu@automattic.com>

georgeh commented:
@jasmussen @shaunandrews This does a little bit with highlighting a Link block that is missing the URL and popping up the input fields. A really rough sketch, but please let me know if it's on the right track. (Setting the URL and text may or may not work)

jasmussen commented:
Excited about this one! 

There's not a ton to see yet:

<img width="351" alt="Screenshot 2021-02-09 at 12 16 08" src="https://user-images.githubusercontent.com/1204802/107356093-a8a41700-6ad0-11eb-9935-2a8abdb1df02.png">

One thing that I think is pretty key for the placeholder menu item, perhaps beyond opening the linkcontrol by default, is that the menu item label is not editable at all until both URL and Label have been filled out (or for Page link, a page has been chosen). That means the menu item is completely inert until configured. I suppose we can keep the RichText component and inert it using `pointer-events: none;` and `cursor: pointer;`, however that might not work as you could probably still tab into it. 

Can we remove the rich text part of the menu item, until configured?

shaunandrews commented on packages/block-library/src/navigation-link/edit.js:
```diff
@@ -372,13 +374,31 @@ function NavigationLinkEdit( {
 							onClose={ () => setIsLinkOpen( false ) }
+							className="wp-block-navigation-link__popover-link-input"
 						>
-							<LinkControl
```
FWIW, I don't think we should stop using LinkControl, or necessarily change LinkControl to resolve #28440. The goal of that issue is really just about the interactions around a Navigation Link block when it doesn't have a link set; Showing the LinkControl when and "empty" Navigation Link is selected.

jasmussen commented:
To Shaun's point, it could make sense to rewind this and make it even simpler:

- Insert menu item and the link control is immediately open with focus there. 
- You can't focus the menu item label, focus is always immediately transferred to the link control.
- Ideally, the menu item is inert until a URL has been added.
- Once a URL has been added, _now_ you can edit the label in the menu item, directly.

Did I get that right?

shaunandrews commented:
Yes, exactly.

georgeh commented:
Thanks! That's was the direction I've been heading in to see what we can do without overhauling too much, but that list works even better than where I was.

Commit:
Use the existing LinkControl for Link Block

Attempting to force focus to the LinkControl when the URL is not set, as
well as highlight a block with a missing URL.

georgeh commented:
OK, is this a bit closer to the interaction that we are trying to hit?

https://user-images.githubusercontent.com/128240/107568717-64d70d80-6bb5-11eb-8b2b-91feaaafd166.mov

shaunandrews commented:
That seems to be on the right path! You can do anything until you select a link. Very nice.

Commit:
Clean up implementation

jasmussen commented:
Really nice work. I think the next step is to polish the visuals a bit. Shaun did you want to try that dimmed gray background from your mockup? Happy to give that a stab and push CSS here. 

Question: from looking at your video, the linkcontrol popover appears top/left aligned from where you click, whereas I believe it's top/center aligned in the master branch. Is this intentional and/or easy to change?

Bonus: can you make the linkcontrol use the `isAlternate` visual treatment? 

```
const POPOVER_PROPS = {
	isAlternate: true,
};
```

That would make it have a black border like the other block UI.

georgeh commented:
> Question: from looking at your video, the linkcontrol popover appears top/left aligned from where you click, whereas I believe it's top/center aligned in the master branch. Is this intentional and/or easy to change?

I think this is a visual artifact because I was using a small viewport to record the video. I just tested it fullscreen and I think it's aligning correctly:

![Screen Shot 2021-02-11 at 11 10 18 AM](https://user-images.githubusercontent.com/128240/107663795-c486f480-6c59-11eb-9bc7-e9f5480e616c.png)

If not, I'm not sure which part needs to move, maybe take a screenshot of the video? (That's also with `isAlternate` on the popover)

Commit:
Add border to popover

georgeh commented:
@jasmussen do you want to take a stab at the visuals for the "Missing URL" state?

jasmussen committed:
Polish

jasmussen commented:
I pushed a little polish:

<img width="768" alt="Screenshot 2021-02-15 at 11 09 10" src="https://user-images.githubusercontent.com/1204802/107932848-4c347200-6f7e-11eb-8c93-d41d9645841d.png">

It now has the default font and size, and a gray background reminiscent of the initial setup state.

@shaunandrews will be asking about the spacing here:

<img width="548" alt="Screenshot 2021-02-15 at 11 09 03" src="https://user-images.githubusercontent.com/1204802/107932974-74bc6c00-6f7e-11eb-91b1-3c44ffbe3c3f.png">

That's the default spacing applied to `wp-block-navigation-link__content`. I the markup, this happens:

<img width="456" alt="Screenshot 2021-02-15 at 11 11 13" src="https://user-images.githubusercontent.com/1204802/107933041-8998ff80-6f7e-11eb-9931-581f14e72d54.png">

I did not find an easy way to make the gray background fit the available space. I'll tinker a bit more.

jasmussen committed:
Get the sizing to work.

jasmussen commented:
Okay, got it working:

<img width="855" alt="Screenshot 2021-02-15 at 11 17 14" src="https://user-images.githubusercontent.com/1204802/107933725-6c186580-6f7f-11eb-9cfb-966fcb12e414.png">

jasmussen commented:
So, in #28440, Shaun designs a "active state" for the placeholder menu item being edited. Like so:

<img width="350" alt="Screenshot 2021-02-15 at 10 56 32" src="https://user-images.githubusercontent.com/1204802/107933878-9d913100-6f7f-11eb-8e42-77e6b7758c26.png">

The inactive state has the gray background:

<img width="486" alt="Screenshot 2021-02-15 at 10 56 27" src="https://user-images.githubusercontent.com/1204802/107933893-a2ee7b80-6f7f-11eb-811e-fd795ce3b967.png">

But we have no such active state:

<img width="872" alt="Screenshot 2021-02-15 at 11 18 35" src="https://user-images.githubusercontent.com/1204802/107933917-ab46b680-6f7f-11eb-8620-d8f89841f63c.png">

Should we have one? 

---

Another thing. From looking at this (https://github.com/WordPress/gutenberg/pull/28861/files#diff-161a9737e181e56e07148b7eee4795f129aa821f73993a276cf1fd9c17941b70R388) there's a click handler to open the link dialog. But this handler doesn't appear to open if I set keyboard focus, and press "Enter", so we probably need a different handler that works with both mouse and keyboard:

<img width="872" alt="Screenshot 2021-02-15 at 11 18 35" src="https://user-images.githubusercontent.com/1204802/107934200-0d072080-6f80-11eb-8b7d-44d5a0dd15e0.png">

---

Finally, there's a conversation on verbiage. In the example here, I have a Page link, a Post link, and a custom Link. You can see the differentiation in menu item:

![menu](https://user-images.githubusercontent.com/1204802/107934395-49d31780-6f80-11eb-868c-98b479b635a3.gif)

In his mockups. Shaun showed different labels for each. George is it feasible, without a great deal of complexity, to show different placeholder text depending on the item type? So instead of "Missing URL", you'd get:

- Page Link: Select a page
- Post Link: Select a post
- Link: Add a link

@shaunandrews I trust you to correct my reading of the above ☝️

jasmussen committed:
Color the separator.

jasmussen commented:
Colorized the link control separator for when the style is "alternate":

<img width="461" alt="Screenshot 2021-02-15 at 11 28 31" src="https://user-images.githubusercontent.com/1204802/107934954-ff05cf80-6f80-11eb-9064-1ed212a59536.png">

Commit:
Customize Missing URL text based on variation

Commit:
Trigger URL Input on Keyboard Focus

Co-authored-by: Kerry Liu <kerry.liu@automattic.com>

georgeh commented on packages/block-editor/src/components/url-input/index.js:
```diff
@@ -100,6 +100,9 @@ class URLInput extends Component {
 	componentWillUnmount() {
+		if ( this.suggestionsRequest?.cancel ) {
+			this.suggestionsRequest.cancel();
+		}
```
Found this during debugging, if you quickly toggle the `URLInput` before the fetch returns, it will attempt to call `setState()` on the unmounted component. This cancels the `suggestionsRequest` Promise when it unmounts to avoid the issue.

I can break it out to a separate PR if needed, the rest of the code doesn't trigger this anymore.

gwwar replied:
Probably okay to sneak in, maybe just leave a quick note in the summary. eg "also fixes ..."

Minor: We can remove the if check using optional chaining.

```
this?.suggestionsRequest?.cancel?.()
```

See also: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining#stacking_the_optional_chaining_operator

Commit:
Update missing labels

Commit:
Add Highlighting to Inactive State

Note that we had to use pretty high specificity to override the text color

Co-authored-by: Kerry Liu <kerry.liu@automattic.com>

georgeh commented:
@jasmussen I worked with @gwwar and we were able to get the chooser menu to trigger on keyboard navigation, add highlighting when the nav menu is not selected, and update the placeholder text. Take a look and let me know what you think.

https://user-images.githubusercontent.com/128240/108095707-28406180-704e-11eb-8d2a-981b6790a55c.mov

georgeh commented:
@shaunandrews Joen's AFK for a few days, if you have time I'd appreciate your thoughts.

shaunandrews commented:
This is coming together nicely!

Here's a few notes:

<img width="487" alt="image" src="https://user-images.githubusercontent.com/191598/108118874-f7235980-706c-11eb-9a50-77f64ea04e87.png">

When I click the empty link block, the block itself is focused. Instead, I'd expect focus to move to the search field within the popover. Also, the link popover seems mis-aligned with the link block; I'd expect them to align to their left edges.

--

When I use my keyboard to select the empty link block, the Link UI immediately opens. This feels wrong. Instead, I'd expect to have to hit `return` to open the Link UI and then have my focus move into the popover.

--

Perhaps unrelated to this PR, but once I move my focus into the Link UI popover I have no way to close the popover. I expect that I could hit `esc` to close the popover and put my focus back on the empty link block, but it doesn't work.

--

<img width="876" alt="image" src="https://user-images.githubusercontent.com/191598/108120289-0b685600-706f-11eb-8091-f1f1ba887b98.png">

The empty state is a little weird in the horizontal variation of the Navigation block, where the width of the link block is much larger than the empty state shape. Perhaps this is expected — and if you put it in a Columns block, its feels much better — I just wanted to note the strangeness.

jasmussen committed:
Polish border and spacing.

jasmussen commented:
I pushed a little fix to the border and the margin:

<img width="807" alt="Screenshot 2021-02-22 at 12 06 59" src="https://user-images.githubusercontent.com/1204802/108700242-9b425000-7506-11eb-9bc3-af3f54e439fb.png">

The focus ring is now correctly spaced all around the placeholder, and the border is inset so it has the same size as the gray solid blob. 

These two:

> When I click the empty link block, the block itself is focused. Instead, I'd expect focus to move to the search field within the popover. Also, the link popover seems mis-aligned with the link block; I'd expect them to align to their left edges.

> When I use my keyboard to select the empty link block, the Link UI immediately opens. This feels wrong. Instead, I'd expect to have to hit return to open the Link UI and then have my focus move into the popover.

Great point. 

- Click with a mouse on the placeholder, and focus is taken to the URL field.
- Select with keyboard purely sets focus on the button. Enter or Space when there, should do the same as a click.

Sorry if I misled on this before.

> The empty state is a little weird in the horizontal variation of the Navigation block, where the width of the link block is much larger than the empty state shape. Perhaps this is expected — and if you put it in a Columns block, its feels much better — I just wanted to note the strangeness.

I'll take a look.

jasmussen commented:
> The empty state is a little weird in the horizontal variation of the Navigation block, where the width of the link block is much larger than the empty state shape. Perhaps this is expected — and if you put it in a Columns block, its feels much better — I just wanted to note the strangeness.

```
.wp-block-navigation.is-vertical wp-block-navigation-link {
	display: inline-block;
}
```

that'll fix it — but I'm tempted to think we should fix this separately, because it's a general issue with the navigation block menu items:

![focus size](https://user-images.githubusercontent.com/1204802/108701382-2b34c980-7508-11eb-8064-aed2ab7de7c6.gif)

And it does kind of open the question: what should the default behavior for that vertical version be? Happy to make a PR to add this separately, let me know!

jasmussen commented:
George, this PR is in good shape! It needs a rebase, and then Shaun's point about focus addressed, then we can land it. Great work!

georgeh commented:
Right now I'm reworking how the focus events happen, which should address Shaun's points. What I'm trying to do is:

Given the URL is not set:
- if you click on the block, it will pop open the URL popover
- if you select the block with the keyboard _and_ then hit return, the URL popover will open
- when the URL popover opens, the text field will get focus
- when you press escape, the URL popover closes

Once that all works I'll get another review and get this merged

Commit:
Add keyboard shortcuts for displaying popover

Commit:
Prettier

georgeh commented:
@jasmussen and @shaunandrews this should be ready to test

jasmussen commented:
This is really nice:

![nav](https://user-images.githubusercontent.com/1204802/108976954-5d6d3500-7688-11eb-8720-bcd1926b3764.gif)

I can't quite put my finger on it, but there appear to be some keyboarding issues, I think it may be general to the navigation block. I'd appreciate a @shaunandrews look on this.

But specifically with mouse usage, focus remains on the button, and is not transferred to the input field. Observe the difference between the appender and the placeholder:

![focus](https://user-images.githubusercontent.com/1204802/108977348-bdfc7200-7688-11eb-9471-5a944b96130f.gif)

Notice how clicking the appender sets focus in the search field, and pressing escape closes it. Clicking the page placeholder does not. What does the appender do here to make this happen, and can we copy it?

georgeh commented:
When I was testing, clicking and keyboarding both triggered focus on the URL input in Firefox DE. I'll test more in Chrome and with the block variations.

shaunandrews commented:
> I can't quite put my finger on it, but there appear to be some keyboarding issues, I think it may be general to the navigation block. I'd appreciate a @shaunandrews look on this.

I mentioned the keyboard focus thing earlier, and can confirm the focus still sticks with the placeholder — I'd expect it to move into the popover and focus the search input.

Related to this, once the popover does appear, its impossible to get rid of it without selecting another block. This is probably not a big deal, but as a keyboard user, I'd expect some way to get _back_ to block itself; I instinctually try hitting `esc` and this just switches to select mode. With other popovers, hitting `esc` closes the popover and puts my focus back to whatever triggered the popover.

I'd expect the placeholder state to work like a toggle. Click it one, and the popover opens and my focus moves to search box. Click it a second time and the popover should close and my focus should move back to the block.

--

Very minor, but I think there should be a `cursor: pointer` applied to the placeholder state. Right now my cursor doesn't change, so its not obvious I can click on the placeholder.

shaunandrews commented on packages/block-library/src/navigation-link/edit.js:
```diff
@@ -389,38 +419,62 @@ export default function NavigationLinkEdit( {
 						<Popover
 							position="bottom center"
 							onClose={ () => setIsLinkOpen( false ) }
+							isAlternate
```
I'm making this same change over in https://github.com/WordPress/gutenberg/pull/28755

I'm not sure its super relevant to this PR, so I question its inclusion here. But also, it seems there might be some discussion around wether this change is good or not.

georgeh replied:
Removed, per Joen

Commit:
Fix Keyboard and Click Focus

georgeh commented:
@shaunandrews can you take another look? There was a regression when I rebased earlier.

Commit:
Add pointer cursor for missing state

jasmussen commented:
This is what I see:

![now](https://user-images.githubusercontent.com/1204802/109110199-f2c70280-7736-11eb-8dbd-0ea504d50b54.gif)

It feels like it's perfect to me now! And it's a _vast_ improvement over what we had before, night and day. This is great.

jasmussen reviewed (approved):
Due to the "is-alternate" feedback from Shaun's PR, let's also remove it here (see https://github.com/WordPress/gutenberg/pull/28861#discussion_r582211556) — we can alway revisit the heuristic for when to show one or the other. So you can remove that, but otherwise, this is great.

If she has time, I'd also like one quick sanity check from @gwwar, but outside of that, this is delicious. Let's land it as soon as we can!

Commit:
Remove isAlternate

gwwar commented on packages/block-library/src/navigation-link/edit.js:
```diff
@@ -332,6 +336,28 @@ export default function NavigationLinkEdit( {
+			missingText = __( 'Select a tag' );
+			break;
+		default:
+			missingText = __( 'Add a link' );
```
I think this default works 👍  If we'd like CPT or custom Taxonomy labels later on, I think we end up needing new post type/taxonomy labels in WP

gwwar replied:
I think we've been leaving translator comments inline so something like the following:
```jsx
/* translators: label for missing values in navigation link block */
missingText = __( 'Add a link' );
```

georgeh replied:
I think that we could make this more extensible in the future by moving the text to `variations.js`, but there are a bunch of other bits we'd want to make extensible as well so for now I prefer this implementation.

gwwar commented:
For folks reviewing adding a `?w=1` will ignore whitespace changes on the diff view. https://github.com/WordPress/gutenberg/pull/28861/files?w=1

gwwar reviewed (approved):
Great stuff @georgeh. This feels like a great improvement! I left a few minor notes, but I think this is pretty good to go. 

https://user-images.githubusercontent.com/1270189/109188311-aa641080-7747-11eb-9ed2-162e7fb53745.mp4

shaunandrews reviewed (approved):
Looks good! Nice work.

Commit:
Simplify Promise cancel

Commit:
Add translator notes

georgeh commented:
TIL you can use `?.()` to optionally call a method, thanks @gwwar 

Clear to merge?

gwwar commented:
Good to go once checks pass ✅
